lib/Runtime/Library/JavascriptString.cpp

```diff
@@ -3885,13 +3894,96 @@ namespace Js
             return false;
         }
 
-        if (wmemcmp(leftString->GetString(), rightString->GetString(), leftString->GetLength()) == 0)
+        LPCWSTR rstr = rightString->GetString();
+        LPCWSTR lstr = leftString->GetString();
+
+        if (lstr == rstr)
         {
             return true;
         }
+
+        if (wmemcmp(lstr, rstr, leftString->GetLength()) == 0)
+        {
+            // at this point we know both leftString and rightString are flattened
+            // why not share the buffer?
+            T::ShareStringBuffer(leftString, rightString);
+            return true;
+        }
         return false;
     }
 
+    void JavascriptString::ShareStringBuffer(JavascriptString * leftString, JavascriptString * rightString)
+    {
+        LPCWSTR rstr = rightString->GetString();
+        LPCWSTR lstr = leftString->GetString();
+
+        Assert(leftString->GetLength() == rightString->GetLength());
```

Comment:
Maybe I should also add an Assert that both left and right strings shouldn't be propertyString?